[Docs] Out of place esimd radix sort documentation update (#1583)
oneapi-src/oneDPL: documentation/library_guide/kernel_templates/esimd/radix_sort.rst

The ``radix_sort`` and ``radix_sort_by_key`` functions sort data using the radix sort algorithm.
The sorting is stable, ensuring the preservation of the relative order of elements with equal keys.

> The functions implement a Onesweep* [#fnote1]_ algorithm variant. Both in place and out of place

Reviewer: No, "overload" looks fine.

> overloads are provided.  For out of place overloads, the input data order is preserved.

Author: fixed

   sycl::event
   radix_sort (sycl::queue q, Range&& r, KernelParam param); // (1)

   template <bool IsAscending = true, std::uint8_t RadixBits = 8,
             typename KernelParam, typename Iterator>
   sycl::event
   radix_sort (sycl::queue q, Iterator first, Iterator last,
               KernelParam param); // (2)

   // Sort a single sequence out of place

   template <bool IsAscending = true, std::uint8_t RadixBits = 8,
             typename KernelParam, typename Range1, typename Range2>
   sycl::event

>    radix_sort(sycl::queue q, Range1&& r, Range2&& r_out,
>               KernelParam param) // (3)

Reviewer: Let's align the calls to follow that suggestion: https://github.com/oneapi-src/oneDPL/pull/1328#discussion_r1466631788.

```suggestion
   radix_sort (sycl::queue q, Range1&& r, Range2&& r_out,
               KernelParam param) // (3)
```

Author: done.

   template <bool IsAscending = true, std::uint8_t RadixBits = 8,
             typename KernelParam, typename Iterator1,
             typename Iterator2>
   sycl::event

>    radix_sort(sycl::queue q, Iterator1 first, Iterator1 last,
>               Iterator2 first_out, KernelParam param) // (4)

Reviewer: ```suggestion
   radix_sort (sycl::queue q, Iterator1 first, Iterator1 last,
               Iterator2 first_out, KernelParam param) // (4)
```

Author: done

   // Sort a sequence of keys and apply the same order to a sequence of values

   template <bool IsAscending = true, std::uint8_t RadixBits = 8,
             typename KernelParam, typename KeysRng, typename ValuesRng>
   sycl::event
   radix_sort_by_key (sycl::queue q, KeysRng&& keys,
                      ValuesRng&& values, KernelParam param); // (5)

   template <bool IsAscending = true, std::uint8_t RadixBits = 8,
             typename KernelParam, typename Iterator1, typename Iterator2>
   sycl::event
   radix_sort_by_key (sycl::queue q, Iterator1 keys_first, Iterator1 keys_last,
                      Iterator2 values_first, KernelParam param); // (6)

   // Sort a sequence of keys and values out of place

   template <bool IsAscending = true, std::uint8_t RadixBits = 8,
             typename KernelParam, typename KeysRng1, typename ValsRng1,
             typename KeysRng2, typename ValsRng2>
   sycl::event

>    radix_sort_by_key(sycl::queue q, KeysRng1&& keys, ValsRng1&& values,
>                      KeysRng2&& keys_out, ValsRng2&& vals_out,
>                      KernelParam param) // (7)

Reviewer: ```suggestion
   radix_sort_by_key (sycl::queue q, KeysRng1&& keys, ValsRng1&& values,
                      KeysRng2&& keys_out, ValsRng2&& vals_out,
                      KernelParam param) // (7)
```

Author: done

+-----------------------------------------------+---------------------------------------------------------------------+
| Name                                          | Description                                                         |
+===============================================+=====================================================================+

> | ``q``                                         | SYCL* queue to submit the kernels to.                               |

Reviewer 2: (Feedback based on #1559)
```suggestion
| ``q``                                         | The SYCL* queue where kernels are submitted.                        |
```

+-----------------------------------------------+---------------------------------------------------------------------+
|                                               |                                                                     |

> |                                               | The sequences of elements to apply the algorithm to.                |

Reviewer 2: (Feedback based on https://github.com/oneapi-src/oneDPL/pull/1559)
```suggestion
|                                               | The sequences to apply the algorithm to.                            |
```

   1 2 3 3 3 5
   s o r t e d

> ``radix_sort`` out of place Example

Reviewer 3: Please change line five as well, to remove the monospace formatting.

> ``radix_sort`` out of place Example

Author: fixed.  removed monospace in a number of headers and capitalized headers.

> ----------------------

Reviewer 3: The header underline needs to go the full length of the header, otherwise it renders incorrectly. Please fix the length.

Author: fixed.

> .. code:: cpp

Reviewer: The examples section became disproportionally large compared to the rest documentation. But still, I do not treat it too large to block the PR, so the suggestions below are optional or/and may be done separately.

There are several things to consider:

- Make the examples collapsible. https://github.com/executablebooks/sphinx-togglebutton may be useful.
- Move the example to the bottom of the page.
- Move the example to the separate files and provide links.

Author: @dcbenito, is there any problem with using the collapsible examples?  I think I like that idea the best.

Reviewer: I am also considering splitting that page in two parts: `radix_sort` and `radix_sort_by_key`. That will solve the issue. 

Let me try it. I will make a branch from this one: it should not take too much time. Then, we can decide if it worth continuing or we should do something different.

Author: I'm experimenting with the toggle stuff, and setting up my environment to be able to do the build locally rather than relying on the artifacts from the runners.
We can compare the two options.

Author: After struggling with getting sphinx togglebutton to render properly with a code block for a while, I don't think it is worth the trouble.  I also tried a custom css solution I found but that also has issues in combination with a code block.  

I'm sure there is a way to do it, but I think splitting the pages is the better option.  

Do you want to work to merge this PR first and then split? close this in favor of your combined PR, or target your PR at this branch?

Reviewer: Let's merge this PR first: this PR is valuable as it is, we will keep the changes isolated.

Reviewer: Meanwhile, I am working on splitting the page in #1586.

Reviewer 3: I was worried that togglebutton wouldn't work. I think splitting the page makes sense.

   // possible build and run commands:
   //    icpx -fsycl radix_sort.cpp -o radix_sort -I /path/to/oneDPL/include && ./radix_sort

   #include <cstdint>
   #include <iostream>
   #include <sycl/sycl.hpp>

   #include <oneapi/dpl/experimental/kernel_templates>

   namespace kt = oneapi::dpl::experimental::kt;

   int main()
   {
      std::size_t n = 6;
      sycl::queue q{sycl::gpu_selector_v};
      std::uint32_t* keys = sycl::malloc_shared<std::uint32_t>(n, q);
      std::uint32_t* keys_out = sycl::malloc_shared<std::uint32_t>(n, q);

      // initialize
      keys[0] = 3, keys[1] = 2, keys[2] = 1, keys[3] = 5, keys[4] = 3, keys[5] = 3;

      // sort
      auto e = kt::gpu::esimd::radix_sort<false, 8>(q, keys, keys + n, keys_out, kt::kernel_param<416, 64>{}); // (4)
      e.wait();

      // print
      for(std::size_t i = 0; i < n; ++i)
         std::cout << keys[i] << ' ';
      std::cout << '\n';
      for(std::size_t i = 0; i < n; ++i)
         std::cout << keys_out[i] << ' ';
      std::cout << '\n';

      sycl::free(keys, q);
      sycl::free(keys_out, q);
      return 0;
   }

**Output:**

.. code:: none

   3 2 1 5 3 3
   5 3 3 3 2 1

> ``radix_sort_by_key`` out of place Example

Reviewer 3: Please change this header to: radix_sort_by_key Out of Place Example 
The header underline needs to go the full length of the header, otherwise it renders incorrectly. Please fix the length.

Author: fixed.
